### PR TITLE
Wip/hughsie/generic guid

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c
Comment (d4s) [inline plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c:494]: Definitely no 0x0.
for GL3224 it is `BLOCK\VEN_05E3&DEV_0749&CID_FFFFFFFF` and for GL3232 -- `BLOCK\VEN_05E3&DEV_0764&CID_22FFFFFF`
Let me clarify numbers with Genesys team.

**Comment by d4s** [comment]
@hughsie did it possible to add the similar behavior for `fu_device_build_instance_id()`? or, better, make the flag not for the whole device but for the selected GUIDs? Rationale: changes similar to 4b38d3f6a417d2317cb98e192481f98f0dca6e5b should have ability to mark that the GUID based on those values is generic.

**Comment by hughsie** [comment]
@d4s that's exactly what `ShowGuid = false` is, no? In that first patch now `BLOCK\VEN_05E3&DEV_0749` won't show up in `fwupdmgr get-devices` at all.

File: libfwupd/fwupd-enums.h
Comment (superm1) [inline libfwupd/fwupd-enums.h:557]: Don't like this name for describing this. It's really preventing the use of matching a guid and forcing other requirements. In those cases aren't we better off having a flag that enforces requirements are set? And/or removing the "generic" guids from problematic devices?

**Comment by d4s** [comment]
@hughsie what about custom extensions like in 4b38d3f6a417d2317cb98e192481f98f0dca6e5b:
`ce88d98c-d36c-5b55-bac4-f07af9e0c022 ← BLOCK\VEN_05E3&DEV_0764&VER_29`

`&VER_29` is also a generic name but used here to distinguish firmware variants.

**Comment by hughsie** [comment]
> &VER_29 is also a generic name but used here to distinguish firmware variants

Is it used as a requirement in the firmware, or just provided in case a quirk might match? if the latter, I think we should do something like:

    -       return fu_device_build_instance_id(FU_DEVICE(self),
    +       return fu_device_build_instance_id_quirk(FU_DEVICE(self),

-- we should also document what GUIDs are *quirk only* in the READMEs I suppose too.

File: libfwupd/fwupd-enums.h
Comment (hughsie) [reply libfwupd/fwupd-enums.h:557]: I agree the flag name is crap too. Maybe `FWUPD_DEVICE_FLAG_ENFORCE_REQUIRES` or something like that?

Comment (superm1) [reply libfwupd/fwupd-enums.h:557]: Yeah that's a better name.

And yeah I was thinking an internal flag instead of an exported flag if it's really just for one plugin. But if it's gonna be used over and over may as well go into global flags

File: plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c
Comment (hughsie) [reply plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c:494]: I've used those now, and lets wait back to hear from Genesys too -- thanks for sending that email.

File: libfwupd/fwupd-enums.h
Comment (hughsie) [reply libfwupd/fwupd-enums.h]: I've moved it to internal flags -- you're right that it's not needed in clients.

**Comment by hughsie** [comment]
Hi all. I'll tag you all here as this pull request might affect existing devices, and I wanted feedback and approval before merging:

 * @MichaelCheng04 --  the elanfp plugin
 * @waynehuang2022 --  the focalfp plugin
 * @jimzhang2 --  the fpc plugin
 * @boger047 --  the goodix-moc plugin
 * @xulinkun --  the goodix-tp plugin

During a security review it's been identified that your devices use a "generic" supplier USB vendor ID, and there is some concern that it's technically possible that OEM A might accidentally deploy firmware updates onto OEM B's hardware.

The best thing to do for future devices is to use the OEM USB vendor ID -- which means we don't have to allow both OEM A and OEM B to update a shared ISV vendor ID. This means that Dell can only update hardware branded as Dell, and Lenovo can only update hardware branded as Lenovo, etc. Most sub-components of things like docks already use the vendor ID, and your devices are the ones that are unusual that they don't :)

For devices already in the field, or that cannot be changed -- the firmware on the LVFS *might* need a small modification: no re-upload is required. We need to **add** a requirement to make sure only the update only applies for the correct OEM, and this can be done using a CHID as documented in https://lvfs.readthedocs.io/en/latest/metainfo.html#using-chid

You can find the CHID for a specific platform by running "sudo fwupdtool hwids" -- e.g. below shows that I'd use `6de5d951-d755-576b-bd09-c5cf66b27234` if I wanted to restrict the update to Lenovo hardware:

	$ sudo fwupdtool hwids
	Computer Information
	--------------------
	BiosVendor: LENOVO
	BiosVersion: N2VET45W (1.30 )
	BiosMajorRelease: 1
	BiosMinorRelease: 30
	FirmwareMajorRelease: 01
	FirmwareMinorRelease: 0b
	Manufacturer: LENOVO
	Family: ThinkPad P1 Gen 3
	ProductName: 20TJS2F44A
	ProductSku: LENOVO_MT_20TJ_BU_Think_FM_ThinkPad P1 Gen 3
	EnclosureKind: a
	BaseboardManufacturer: LENOVO
	BaseboardProduct: 20TJS2F44A
	Hardware IDs
	------------
	{687b4b40-14cf-5a9e-b9a2-84d93a5f99c9}   <- Manufacturer + Family + ProductName + ProductSku + BiosVendor + BiosVersion + BiosMajorRelease + BiosMinorRelease
	{d6dffdd4-f8c5-56b2-8a54-29d4554a81b7}   <- Manufacturer + Family + ProductName + BiosVendor + BiosVersion + BiosMajorRelease + BiosMinorRelease
	{2f6d9243-39b3-561b-acb5-33c664ac656e}   <- Manufacturer + ProductName + BiosVendor + BiosVersion + BiosMajorRelease + BiosMinorRelease
	{bf5fea93-df40-5336-922f-a2fa3ae67e5f}   <- Manufacturer + Family + ProductName + ProductSku + BaseboardManufacturer + BaseboardProduct
	{5ac99727-ff1a-5d8a-aef4-8a03d8f5cfe3}   <- Manufacturer + Family + ProductName + ProductSku
	{742d6de4-82fc-56b8-8079-963759c68dea}   <- Manufacturer + Family + ProductName
	{ff4db33f-0e86-52e0-b531-5ecf58c194c1}   <- Manufacturer + ProductSku + BaseboardManufacturer + BaseboardProduct
	{3969d15b-9209-5350-9f13-966ee621eb4b}   <- Manufacturer + ProductSku
	{a4d618e0-6988-5177-800c-64405dbbcefa}   <- Manufacturer + ProductName + BaseboardManufacturer + BaseboardProduct
	{2dc64c9e-97c4-59bf-b0be-892546d7e655}   <- Manufacturer + ProductName
	{536c0c4c-e739-5bc7-8f93-042c73d961ef}   <- Manufacturer + Family + BaseboardManufacturer + BaseboardProduct
	{3667f735-639c-5ca5-9ae7-a464d4b49e27}   <- Manufacturer + Family
	{e093d715-70f7-51f4-b6c8-b4a7e31def85}   <- Manufacturer + EnclosureKind
	{eb8b729b-355c-5144-b23c-90f24e542a8c}   <- Manufacturer + BaseboardManufacturer + BaseboardProduct
	{6de5d951-d755-576b-bd09-c5cf66b27234}   <- Manufacturer

This can be added on the LVFS like below:

Screenshot 2023-11-04 at 12-39-09 LVFS Firmware Component Details

If you need more permission to change a firmware in testing/stable please give me details as I'm an admin and can do it for you. I'm happy to answer questions either here or my private email. Thanks!

File: libfwupdplugin/fu-device.h
Comment (d4s) [inline libfwupdplugin/fu-device.h:585]: nit: 1ull ?

Comment (hughsie) [reply libfwupdplugin/fu-device.h:585]: Ohh, indeed. I'm not 100% which is the correct way around -- I think we use both in the codebase!

Comment (hughsie) [reply libfwupdplugin/fu-device.h:585]: **Both are allowed by the C standard** --  (section 6.4.4.1)

File: libfwupdplugin/fu-device.h
Comment (superm1) [inline libfwupdplugin/fu-device.h:596]: why did you swap this to `llu` instead of `ull`?  All the others have `ull`

File: src/fu-engine.c
Comment (superm1) [inline src/fu-engine.c:2221]: it could also be a CHID couldn't it be?

File: src/fu-engine.c
Comment (superm1) [inline src/fu-engine.c:2216]: do we actually want to have a `--force` here?  I'd think that's a way to defeat security as `--force` can be added by the client.

Comment (superm1) [reply src/fu-engine.c:2216]: Maybe only in `SUPPORTED_BUILD=0`?

**Review by superm1 — DISMISSED** [review]
I'm mostly fine with this, just some really small stuff to change or waive.

File: libfwupdplugin/fu-device.h
Comment (hughsie) [reply libfwupdplugin/fu-device.h:596]: Ohh, it was originally in libfwupd which confusingly has the other one. **Choose one** and I'll convert the codebase in another PR. :)

File: src/fu-engine.c
Comment (hughsie) [reply src/fu-engine.c:2221]: Agreed, thanks. Fixed.

File: src/fu-engine.c
Comment (hughsie) [reply src/fu-engine.c:2216]: I think `SUPPORTED_BUILD` is a great idea, fixed.

File: libfwupdplugin/fu-device.h
Comment (superm1) [reply libfwupdplugin/fu-device.h:596]: Hah. I don't care, just make it consistent ;)

**Review by superm1 — APPROVED** [review]

**Pull request merged**
